spec/consensus/proposer-based-timestamp/pbts_001_draft.md

```diff
@@ -0,0 +1,268 @@
+ } else {
+  schedule OnTimeoutPropose(h_p,round_p) to 
+   be executed after timeoutPropose(round_p)
+ }
+}
+```
+
+</td>
+
+<td>
+
+```go
+function StartRound(round) {
+ round_p ← round
+ step_p ← propose
+ if proposer(h_p, round_p) = p {
+  // new wait condition
+  wait until now_p > block time of block h_p - 1
+  if validValue_p != nil {
+   // add "now_p"
+   proposal ← (validValue_p, now_p) 
+  } else {
+   // add "now_p"
+   proposal ← (getValue(), now_p) 
+  }
+  broadcast ⟨PROPOSAL, h_p, round_p, proposal, validRound_p⟩
+ } else {
+  schedule OnTimeoutPropose(h_p,round_p) to 
+   be executed after timeoutPropose(round_p)
+ }
+}
+```
+
+</td>
+</tr>
+</table>
+
+- Rule on lines 28-35
+
+<table>
+<tr>
+<th>arXiv paper</th>
+<th>Proposer-based time</th>
+</tr>
+
+<tr>
+<td>
+
+```go
+upon timely(⟨PROPOSAL, h_p, round_p, v, vr⟩) 
+ from proposer(h_p, round_p)
+ AND 2f + 1 ⟨PREVOTE, h_p, vr, id(v)⟩ 
+while step_p = propose ∧ (vr ≥ 0 ∧ vr < round_p) do {
+ if valid(v) ∧ (lockedRound_p ≤ vr ∨ lockedValue_p = v) {
+  
+  broadcast ⟨PREVOTE, h_p, round_p, id(v)⟩
+ } else {
+  broadcast ⟨PREVOTE, hp, round_p, nil⟩
+ }
+}
+```
+
+</td>
+
+<td>
+
+```go
+upon timely(⟨PROPOSAL, h_p, round_p, (v, tprop), vr⟩) 
+ from proposer(h_p, round_p) 
+ AND 2f + 1 ⟨PREVOTE, h_p, vr, id(v, tvote)⟩ 
+ while step_p = propose ∧ (vr ≥ 0 ∧ vr < round_p) do {
+  if valid(v) ∧ (lockedRound_p ≤ vr ∨ lockedValue_p = v) {
+   // send hash of v and tprop in PREVOTE message
+   broadcast ⟨PREVOTE, h_p, round_p, id(v, tprop)⟩
+  } else {
+   broadcast ⟨PREVOTE, hp, round_p, nil⟩
+  }
+ }
+```
+
+</td>
+</tr>
+</table>
+
+- Rule on lines 49-54
+
+<table>
+<tr>
+<th>arXiv paper</th>
+<th>Proposer-based time</th>
+</tr>
+
+<tr>
+<td>
+
+```go
+upon ⟨PROPOSAL, h_p, r, v, ∗⟩ from proposer(h_p, r) 
+ AND 2f + 1 ⟨PRECOMMIT, h_p, r, id(v)⟩ 
+ while decisionp[h_p] = nil do {
+  if valid(v) {
+
+   decision_p [h_p] = v
+   h_p ← h_p + 1
+   reset lockedRound_p , lockedValue_p, validRound_p and 
+    validValue_p to initial values and empty message log 
+   StartRound(0)
+  }
+ }
+```
+
+</td>
+
+<td>
+
+```go
+upon ⟨PROPOSAL, h_p, r, (v,t), ∗⟩ from proposer(h_p, r) 
+ AND 2f + 1 ⟨PRECOMMIT, h_p, r, id(v,t)⟩
+ while decisionp[h_p] = nil do {
+  if valid(v) {
+   // decide on time too
+   decision_p [h_p] = (v,t) 
+   h_p ← h_p + 1
+   reset lockedRound_p , lockedValue_p, validRound_p and 
+    validValue_p to initial values and empty message log 
+   StartRound(0)
+  }
+ }
+```
+
+</td>
+</tr>
+</table>
+
+- Other rules are extended in a similar way, or remain unchanged
+
+### Property Overview
+
+#### Safety and Liveness
+
+For safety (Point 1, Point 2, Point 3i) and liveness (Point 4) we need
+the following assumptions:
+
+- There exists a system parameter `PRECISION` such that for any two correct validators `V` and `W`, and at any real-time `t`, their local times `C_V(t)` and `C_W(t)` differ by less than `PRECISION` time units, 
```

Comment:
When you say system parameter, does this imply that `PRECISION` is a consensus parameter that is agreed upon in consensus or an individual node parameter?

Reply:
Same q for `ACCURACY`

Reply:
`PRECISION` needs to be agreed upon in consensus, as it will be part of the code. Moreover, in our analysis we assume it is the same on every node.

`ACCURACY` is not visible in the code. But it is crucial to relate `blocktime` to `real-time`. It required, e.g., by a lightclient to understand how old a block really is. In principle here a lightclient may subjectively say, "I assume accuracy is 10 sec". Then its safety depends on the correctness of this assumption. It would therefore be good if the validators would provide an estimate how large their `ACCURACY` is.

Reply:
Right. We could also persist `real-time` values for each node when they store their blocks and add this as part of an RPC endpoint. Thus, participants of the network could query nodes for their `ACCURACY`. Would doing something like that makes sense?

Reply:
I am not sure that would help. 

But you are right: there is a bigger question: how are validators synchronizing their clocks now (I guess they use NTP) and what values for `ACCURACY` can we assume from there.

Or even more generally: right now there are a lot of implicit assumptions about time and clocks. We need to make them more explicit (as we tried in this PR). And once we made these assumptions explicit, what are suitable values for the parameters. For instance, the timeout values of Tendermint consensus implementation work pretty well. I wonder how these values where set/estimated initially. A similar approach we should take for the parameters introduced in this PR.

Also, I think we can be quite conservative here. Even if the precision/accuracy of the clocks with NTP is in the order of several hundred microseconds, perhaps it is enough to set the parameters in the code to 10 seconds. After all, given that unbonding and trusting periods are in the order of weeks, we just need to maintain some relation of real-time to block time.

The point about these parameters are not to be as precise as possible, but to understand "how precise" the solution guarantees to be in order to have sound guarantees at the system interfaces, e.g., between full nodes and light nodes. The point is to guarantee some value of precision that is sufficient for the application purpose rather to expose the actual precision that the system could achieve.